go/libkb/util.go

```diff
@@ -1210,3 +1210,55 @@ func JsonwStringArray(a []string) *jsonw.Wrapper {
 	}
 	return aj
 }
+
+var throttleBatchClock clockwork.Clock = clockwork.NewRealClock()
+
+type throttleBatchEmpty struct{}
+
+func isEmptyThrottleData(arg interface{}) bool {
+	_, ok := arg.(throttleBatchEmpty)
+	return ok
+}
+
+func ThrottleBatch(f func(interface{}), batcher func(interface{}, interface{}) interface{},
+	reset func() interface{}, delay time.Duration) (func(interface{}), func()) {
+	var lock sync.Mutex
+	var lastCalled time.Time
+	var creation func(interface{})
+	hasStored := false
+	scheduled := false
+	stored := reset()
+	cancelCh := make(chan struct{})
+	creation = func(arg interface{}) {
+		lock.Lock()
+		defer lock.Unlock()
+		elapsed := throttleBatchClock.Since(lastCalled)
+		isEmpty := isEmptyThrottleData(arg)
+		if !isEmpty {
+			stored = batcher(stored, arg)
+			hasStored = true
+		}
+		if elapsed > delay && (!isEmpty || hasStored) {
+			f(stored)
+			stored = reset()
+			hasStored = false
+			lastCalled = throttleBatchClock.Now()
+		} else if !scheduled && !isEmpty {
+			scheduled = true
+			go func() {
+				select {
+				case <-throttleBatchClock.After(delay - elapsed):
+					lock.Lock()
+					scheduled = false
+					lock.Unlock()
+					creation(throttleBatchEmpty{})
+				case <-cancelCh:
+					return
+				}
+			}()
+		}
+	}
+	return creation, func() {
+		close(cancelCh)
```

Comment:
scared this may be called twice and panic

Reply:
if you add more state around this, can throttler just become a struct? feels more intuitive that way

Reply:
Made it just do anything on first call

Reply:
why not `sync.Once`?

Reply:
I forgot it existed